misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterService.kt

```diff
@@ -0,0 +1,210 @@
+package misk.clustering.kubernetes
+
+import com.google.common.util.concurrent.AbstractIdleService
+import com.google.inject.Key
+import io.kubernetes.client.models.V1Pod
+import io.kubernetes.client.util.Watch
+import misk.DependentService
+import misk.clustering.Cluster
+import misk.clustering.ClusterHashRing
+import misk.clustering.ClusterWatch
+import misk.logging.getLogger
+import org.eclipse.jetty.util.BlockingArrayQueue
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicReference
+import javax.inject.Inject
+import javax.inject.Singleton
+import kotlin.concurrent.thread
+
+/**
+ * [KubernetesClusterService] implements the [Cluster] interface on top of the kubernetes APIs.
+ * To make use of this service, pods must run as a service account that has pod list/watch/get
+ * access within its own namespace
+ */
+@Singleton
+internal class KubernetesClusterService @Inject internal constructor(
+  config: KubernetesConfig
+) : AbstractIdleService(), Cluster, DependentService {
+  private val peersRef = AtomicReference<Set<Cluster.Member>>(setOf())
+  private val hashRingRef = AtomicReference<ClusterHashRing>(ClusterHashRing(setOf()))
+  private val running = AtomicBoolean(false)
+  private val events = BlockingArrayQueue<Event>()
+
+  override val hashRing: ClusterHashRing get() = hashRingRef.get()
+  override val peers: Set<Cluster.Member> get() = peersRef.get()
+  override val self: Cluster.Member = Cluster.Member(config.my_pod_name, config.my_pod_ip)
+
+  override val consumedKeys: Set<Key<*>> = setOf()
+  override val producedKeys: Set<Key<*>> = setOf(
+      Key.get(Cluster::class.java),
+      Key.get(KubernetesClusterService::class.java)
+  )
+
+  override fun startUp() {
+    log.info { "starting k8s cluster manager" }
+    running.set(true)
+
+    thread(name = "k8s-cluster-event-loop") {
+      runEventLoop()
```

Comment:
I think you wanna use a AbstractExecutionThreadService and override run()

Reply:
https://github.com/google/guava/wiki/ServiceExplained#abstractexecutionthreadservice

Reply:
oh neat, ok